server/cache.go

```diff
@@ -101,32 +102,99 @@ func (s *storesInfo) getStoreCount() int {
 	return len(s.stores)
 }
 
+// regionMap wraps a map[uint64]*regionInfo and supports randomly pick a region.
+type regionMap struct {
+	m   map[uint64]*regionEntry
+	ids []uint64
+}
+
+type regionEntry struct {
+	*regionInfo
+	pos int
+}
+
+func newRegionMap() *regionMap {
+	return &regionMap{
+		m: make(map[uint64]*regionEntry),
+	}
+}
+
+func (m *regionMap) Len() int {
+	if m == nil {
```

Comment:
I think you can remove the check here because we will let user use newRegionMap to create the region map and the inner map can't be nil.